docarray/array/array/pushpull/s3.py

```diff
@@ -0,0 +1,117 @@
+from typing import Dict, Iterator, List, Optional, Type
+
+import boto3
+import botocore
+from smart_open import open
+from typing_extensions import TYPE_CHECKING
+
+from docarray.array.array.io import BinaryIOLike
+from docarray.array.array.pushpull.helpers import (
+    _BufferedCachingReader,
+    _from_binary_stream,
+    _to_binary_stream,
+)
+
+if TYPE_CHECKING:  # pragma: no cover
+    from docarray import BaseDocument, DocumentArray
+
+
+class PushPullS3:
+    """Class to push and pull DocumentArray to and from S3."""
+
+    @staticmethod
+    def list(namespace: str, show_table: bool = False) -> List[str]:
+        bucket, namespace = namespace.split('/', 1)
+        s3 = boto3.resource('s3')
+        bucket = s3.Bucket(bucket)
+        da_files = [
+            obj.key
+            for obj in bucket.objects.all()
+            if obj.key.startswith(namespace) and obj.key.endswith('.da')
+        ]
+        da_names = [f.split('/')[-1].split('.')[0] for f in da_files]
+        return da_names
+
+    @staticmethod
+    def delete(name: str, missing_ok: bool = True) -> bool:
+        bucket, name = name.split('/', 1)
+        s3 = boto3.resource('s3')
+        object = s3.Object(bucket, name + '.da')
+        try:
+            object.load()
+        except botocore.exceptions.ClientError as e:
```

Comment:
Why is there additional check for 404 here? The boto client already handles this case so why the duplication?

Reply:
Because the response from [object.delete()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/object/load.html) does not indicate if the delete was successful. It returns 204 regardless of whether or not the object we wished to delete existed in the first place.

The [object.load()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/object/load.html) here actually calls [head_object()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/head_object.html) which only retrieves metadata. So it shouldn't be that costly. We aren't paying double cost here.